3주차 과제 - Todo 리스트 테스트 작성하기

sejineeee opened the pull request:
## 내용
안녕하세요. 3일만에 첫 PR을 남겨봅니다...😭
Jest와 React testing library에 대해서 처음 접하다 보니, 첫 PR을 생성하는데 많은 시간이 걸렸습니다. 학습하는  곳에 있던 아샬님의 TDD 영상을 보면서, 되게 직관적인 단어들을 사용하기 때문에 금방 이해할 수 있을거라고 생각했는데... 너무 어려웠어요...ㅠㅠ

FE conf 2020에서 TDD 라이브 코딩을 해주는 영상을 보면서 작성한 것입니다. 아직 혼자서 작성하기에는 많은 어려움이 있습니다. 🙇

## 질문
- 혹시 제가 참고할 수 있는 사이트가 있으면 추천해주시면 감사하겠습니다!
- BDD를 적용하면 `context, describe, it`을 사용한다고 보았습니다. `describe`는 테스트를 그룹화, it은 test의 별칭이라는 것을 검색해서 알게 되었는데, context는 찾아도 나오지 않더라고요. (제가 구글링 능력이 부족해서 그런 걸 수도 있습니다.) plugin-context에 대해서도 보긴 했는데, 아직 이해가 잘 되지 않습니다 🥺
- 뭘 모르는지 잘 몰라서 (무의식적 무지 상태), 어떤 질문을 남겨야 저에게 도움이 될 지 잘 모르겠습니다


[이번 한 주도 열심히 해보도록 하겠습니다. 파이팅!! 🤜]

sejineeee committed:
Page test 파일 구현

sejineeee committed:
List test 파일 구현

sejineeee committed:
Input test 파일 구현

sejineeee committed:
App test 파일 구현

daadaadaah commented on src/App.test.jsx:
```diff
@@ -0,0 +1,8 @@
+
+import App from './App';
+
+test('todo app', () => {
```
연습차원에서 되도록 test 보다 describe-(context)-it 으로 작성해보도록 연습해보세요!ㅎㅎ

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,43 @@
+describe('List', () => {
+  const handleClick = jest.fn();
+
+  const renderList = (tasks) => render(<List tasks={tasks} onClickDelete={handleClick} />);
```
중괄호를 생략할 수 있더라도, 중괄호를 해주는게 저는 가독성이 더 좋더라구요!ㅎㅎprops도 2개이이상이면 세로 정렬로 해주는 것도 좋구!ㅎㅎ

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,43 @@
+
+  const renderList = (tasks) => render(<List tasks={tasks} onClickDelete={handleClick} />);
+
+  context('with tasks', () => {
```
context 를 적절한 상황에서 잘 쓰셨어요!👏👏👏

context는 아래 링크 참고해보세요!
참고 : https://www.betterspecs.org/#contexts

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,43 @@
+      { id: 2, title: '블로그 작성하기' },
+    ];
+
+    it('todo list', () => {
```
테스트 시나리오를 작성하실 경우, describe-context-it 에 있는 문장들이 하나로 합쳐졌을 떄, 한 문장처럼 자연스럽게 읽히도록 작성해보세요!
지금 작성해주신 대로 문장을 만들어보면, `List todo list`로 문장이 만들어져서 어색한 것 같아요! List는 todo list를 어떻게 하는지에 대해 생각해보시면 좋을 것 같아요!ㅎㅎ(다른 곳도 함께 수정해주세요!😊)

daadaadaah commented on src/Page.test.jsx:
```diff
@@ -0,0 +1,30 @@
+
+import Page from './Page';
+
+test('todo page', () => {
+  const tasks = [
+    { id: 1, title: '리액트 공부하기' },
+    { id: 2, title: '블로그 작성하기' },
+  ];
+  const taskTitle = '';
+
+  const handleChangeTitle = jest.fn();
+  const handleClickAddTask = jest.fn();
+  const handleClickDeleteTask = jest.fn();
+
+  const { container, getAllByText } = render((<Page
+    tasks={tasks}
+    taskTitle={taskTitle}
+    onChangeTitle={handleChangeTitle}
+    onClickAddTask={handleClickAddTask}
+    onClickDeleteTask={handleClickDeleteTask}
+  />));
+
+  expect(container).toHaveTextContent('리액트 공부하기');
+  expect(container).toHaveTextContent('블로그 작성하기');
+
+  const buttons = getAllByText('완료');
+
+  fireEvent.click(buttons[0]);
+});
```
이 테스트로 무엇을 기대하는 테스트인지 describe-(context)-it 문으로 표현해봅시다

daadaadaah reviewed (changes requested):
3일만이라도, 비교적 빠르게 PR 날리신 점 너무 잘하셨어요!ㅎㅎ
무지를 드러내주셔야 저도 도움을 드릴 수 있어요!ㅎㅎ질문도 잘 남겨주시고~ 잘하셨습니다!👏

> 혹시 제가 참고할 수 있는 사이트가 있으면 추천해주시면 감사하겠습니다!

- 도움될만한 링크 남겨놨습니다!ㅎㅎ

참고
- https://johngrib.github.io/wiki/junit5-nested/#describe---context---it-%ED%8C%A8%ED%84%B4
- https://marocchino.net/2016/12/04/about-rspec/
- https://ko.javascript.info/testing-mocha#ref-888
- https://github.com/ahastudio/til/blob/main/ruby/20161206-rspec-let.md
- [Ginkgo - Go 언어 개발자를 위한 BDD 테스팅 프레임워크](https://www.youtube.com/watch?v=gfTsSBRvdqI)

> context 관련 자료
- 코멘트에 남겨놨어요!ㅎㅎ

> 뭘 모르는지 잘 몰라서 (무의식적 무지 상태), 어떤 질문을 남겨야 저에게 도움이 될 지 잘 모르겠습니다
- 지금 남겨주신 질문도 잘 하신거예요!👍
- 저는 크게 테스트 코드 작성할 때, 크게 2단계로 생각해요. 
1) 무엇을 테스트 할지 (-> 관련해서 도움될만한 리뷰 링크 남겨두었습니다!)
- https://github.com/CodeSoom/react-week3-assignment-1/pull/134#pullrequestreview-1013694078
- https://github.com/CodeSoom/react-week3-assignment-1/pull/138#pullrequestreview-1015192093
2) 1번을 테스트 코드로 어떻게 작성할지 (-> 이부분은 jest, React-testing-libarary 공식 문서 참고하세요!)
- 어느 단계에서 감이 안오는지 한번 고민해보세요!ㅎㅎ

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,43 @@
+  const renderList = (tasks) => render(<List tasks={tasks} onClickDelete={handleClick} />);
+
+  context('with tasks', () => {
+    const tasks = [
+      { id: 1, title: '리액트 공부하기' },
+      { id: 2, title: '블로그 작성하기' },
+    ];
```
지금 `tasks`가 여러 테스트에서 사용되고 있어요! 여러 곳에서 사용되는 tasks 데스트용 데이터를 보통 fixture라고 이야기합니다. 저는 이러한 데이터를 `fixtures` 폴더 하나를 만들어서 따로 관리합니다!ㅎㅎ

daadaadaah commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,17 @@
+
+  fireEvent.click(getByText('추가'));
+
+  expect(handleChange).not.toBeCalled();
```
저라면, handleChange가 호출되지 않는 것보다 handleClick이 호출된 걸 확인하는 테스트를 작성할 것 같아요!

sejineeee committed:
review: Input test 파일 작성

BDD 적용(describe, it 키워드 사용)
중복되는 render 함수 renderInput 변수에 할당
handleChange, handleClick 테스트 추가
input과 button 렌더링 테스트 추가

sejineeee committed:
feat: tasks 파일 생성

task를 담은 배열 tasks 분리하여 tasks 파일 생성

sejineeee committed:
review: List test 파일 수정

tasks 파일 import
it 키워드 name에 영어로 작성한 부분 한글로 변경
shadow로 인해서 매개변수 및 변수명 변경

sejineeee committed:
review: Page test 파일 수정

BDD 적용(describe, it 키워드 적용)
render 함수 renderPage 변수에 할당
분리한 tasks 파일 import

sejineeee committed:
review: App test 파일 수정

BDD 적용(describe, it 키워드 적용)
handleClick, hnadlChange 함수 테스트
App 렌더링 테스트

sejineeee committed:
console 출력 문구 삭제

sejineeee commented:
## 내용
- render 함수를 분리하며, 화살표 함수로 작성하였을 때, return문에 중괄호를 작성하는 것이 가독성면에서 더 좋다고 해주신 리뷰대로 수정하려고 했으나, eslint를 실행하였을 때, 중괄호가 없이 수정되어 화살표 함수에서 `function` 키워드를 이용하는 것으로 수정하였습니다.
- fixtures 폴더를 생성하고, `tasks.js` 파일을 생성하여 필요한 부분에서 import 해서 사용할 수 있도록 수정하였습니다.

daadaadaah commented on src/App.test.jsx:
```diff
@@ -1,8 +1,50 @@
+    expect(handleClick).toBeCalled();
+  });
+
+  it('handleChange 함수가 실행된다', () => {
```
어떨때, handleChange 함수가 실행되는지 같이 기재해주시면 좋을 것 같아요!ㅎㅎ

daadaadaah reviewed (changes requested):
> App test 파일에서는 렌더링 되는 것만 확인하면 될 거라고 생각했는데, test coverage를 확인했을 때는 handleChange, handleClick, handleClickDelete 함수를 테스트 해야 된다고 해서 이렇게 작성하였으나 아래와 같은 에러가 발생해서 어떻게 해줘야 좋을 지 잘 모르겠습니다!
- App.test 파일에서 렌더링 되는 것만 확인하면 될 텐대, 이따 저녁때 sejineeee님 코드 clone 받아봐서 한번 체크해볼게요!

> input test 파일에서 input의 value가 바뀌는지 확인을 하고 싶었습니다. 그래서 expect(input.value).toBe('TDD')로 테스트 해보려고 했으나, 초기 설정해놓은 "운동하기" 값이 아니어서 에러가 발생합니다. 제가 expect 구문을 잘못 작성한 것일지 아니면 fireEvent.change 구문 사용법을 잘못 알고 작성한 것인지 궁금합니다!!
- Input 컴포넌트는 사용자가 입력한 값 대로 바뀐 걸 확인하지 못합니다! 그저 handlechange가 실행 되는지만 확인할 수 있어요! 값을 변경시키는 것이 누구의 역할인지 한번 다시 생각해봅시다!ㅎㅎ

> const handleChange = jest.fn()을 describe 바깥에서 선언하는게 더 좋을지. 아니면 describe 안에서 선언하는 게 더 좋을지 궁금합니다!!
- 저는 describe('컴포넌트 이름') 밑에 선언해두는 편입니다!

sejineeee committed:
review: 함수 test 수정

input에 입력할 경우, handleChange 함수 test 수정
추가 버튼을 누를 경우, handleClick 함수 test 수정
완료 버튼을 누를 경우, handleClickDelete 함수 test 수정

sejineeee commented:
[a32cb6c](https://github.com/CodeSoom/react-week3-assignment-1/pull/156/commits/a32cb6c8c5a5d1e9e843020f80810ce3269f23b1)
안녕하세요!! 리뷰해주신 것을 토대로 `App.test.jsx` 파일을 수정해보았습니다.
바뀐 값을 확인해주고 싶어서 `input`의 `value`를 확인하였습니다.
App 파일에서 렌더링 하는 부분에서 expect(jest.fn()) 이렇게 들어왔던 부분을 수정하였습니다!
App 파일에서 그냥 보이는 것들만 테스트를 해주려고 했었는데, `coverage`를 확인해본 결과 100 달성을 하지 못하길래 다시 수정해보았습니다! 혹시 제가 작성한 부분에서 수정해야 될 부분이 있거나, 더 나은 방법이 있을 경우 코멘트 남겨주시면 감사하겠습니다!!!

이번 한 주도 감사합니다!! 🎃

daadaadaah commented on src/App.test.jsx:
```diff
@@ -0,0 +1,61 @@
+    expect(container).toHaveTextContent('추가');
+  });
+
+  it('입력하면 handleChange 함수가 실행된다.', () => {
```
App 컴포넌트의 관심사가 무엇인지에 대해 한번 생각해봅시당!
handleChange 함수 잘 호출되어서 어떻게 되는지! 테스트 코드는 잘 작성해주셨는데, 그에 맞게 테스트 이름이 잘 작성되지 않아 아쉬워요😭 (밑에 다른 테스트도 마찬가지구요!)

daadaadaah reviewed (changes requested):
> 아샬님 과제풀이 강의를 보았을 때, expect(getByText('할 일이 없어요')).not.toBeNull(); 이렇게 toBeNull()을 자주 사용하시던데, toHaveTextContent 보다 toBeNull()을 이용하여 테스트를 하는 게 더 권장(추천)하는 방식인가요?
- 같은 결과에 대해 표현하는 방법이 다양한데, 어떤 의도를 좀더 드러내고 싶은지 여부에 따라 다를 것 같아요! toBeNull, toHaveTextContent 둘이 어떤 차이가 있는지에 대해 공부해보시는 걸 추천드려요! 더불어 다른 matcher들도 공부해보시면서 다양한 matcher들을 왜 써야 하는지에 대해 한번 공부해보세요!

> buttons 배열의 인덱스 0번을 넘겨줬기 때문에 toBeCalledWith에도 0을 넘겨줘야 될 거라고 생각했는데, 1이 들어오는 이유가 궁금합니다!! 인덱스 0번의 버튼을 클릭한 거니까 인덱스 0번의 버튼이 호출된 거 아닌가요? 🤔
- toBeCalledWith의 매개변수가 어떤 의미인지에 대해 공부해보세요! 인덱스는 아니거든용!
- 그리고, 그 값이 sejineeee님 코드에서 어떤 값을 갖는지 한번 체크해보세요!

sejineeee committed:
review: test 이름 변경

각 함수를 테스트 하는 부분의 이름을 좀 더 명확하게 변경해보았습니다.

sejineeee commented:
[ee5083c](https://github.com/CodeSoom/react-week3-assignment-1/pull/156/commits/ee5083c180a26a20ebeea150d086dc4490dfe0ad)
리뷰를 토대로 각 함수를 테스트 해주는 부분의 이름을 좀 더 명확하게 작성하는 방향으로 변경했습니다!!
제가 맞게 변경한 건지 확신이 들지 않네요 😢

코멘트 달아주신 내용에 대해서는 공부해보도록 하겠습니다!!!! 🤓 이번 한 주도 감사합니다!!!

daadaadaah reviewed (approved):
정말 구체적으로 적어주셨네용ㅎㅎ
그냥, `할일을 입력하면, 입력한 할일이 보인다.` 등으로 결과만 적어주셔도 괜찮습니다!ㅎㅎ
이번주 고생 많으셨어요~ 다음주에도 같이 힘내봐요!👊
